File: go/internal/csvdiff/profiles.go
package csvdiff

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// Profile is one named comparison from csvdiff.toml, so a recurring comparison
// does not get retyped. Fields are pointers because "absent" and "false"/"0"
// mean different things: only the keys actually present override the defaults.
//
//	[profiles.orders]
//	key       = ["order_id", "line_no"]
//	compare   = ["qty", "price", "status"]   # omit for all common non-key columns
//	ignore    = ["updated_at"]
//	trim      = true
//	tolerance = 0.005
type Profile struct {
	Key         []string
	Compare     []string
	Ignore      []string
	Trim        *bool
	IgnoreCase  *bool
	EmptyIsNull *bool
	Tolerance   *float64
	MaxRows     *int
	Delimiter   *string
	Encoding    *string
	Engine      *string
}

// SearchPath is where a config file is looked for when no path is given.
func SearchPath() []string {
	paths := []string{"csvdiff.toml"}
	if home, err := os.UserHomeDir(); err == nil {
		paths = append(paths, filepath.Join(home, ".config", "csvdiff", "csvdiff.toml"))
	}
	return paths
}

// LoadProfiles reads the profile tables from a config file, or returns an empty
// map when there is none.
//
// This reads the subset of TOML the config format actually uses — [profiles.x]
// tables of scalars and string arrays — rather than pulling in a parser for a
// file that is a dozen lines long.
func LoadProfiles(explicit string) (map[string]Profile, error) {
	candidates := SearchPath()
	if explicit != "" {
		candidates = []string{explicit}
	}
	for _, p := range candidates {
		st, err := os.Stat(p)
		if err != nil || st.IsDir() {
			continue
		}
		f, err := os.Open(p)
		if err != nil {
			return nil, fmt.Errorf("cannot read %s: %w", p, err)
		}
		defer f.Close()
		profiles, err := parseProfiles(f)
		if err != nil {
			return nil, fmt.Errorf("cannot parse %s: %w", p, err)
		}
		return profiles, nil
	}
	if explicit != "" {
		return nil, fmt.Errorf("config file not found: %s", explicit)
	}
	return map[string]Profile{}, nil
}

func parseProfiles(r interface{ Read([]byte) (int, error) }) (map[string]Profile, error) {
	out := map[string]Profile{}
	current := ""
	sc := bufio.NewScanner(r)
	sc.Buffer(make([]byte, 0, 64*1024), 1<<20)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if strings.HasPrefix(line, "[") && strings.HasSuffix(line, "]") {
			name := strings.TrimSpace(line[1 : len(line)-1])
			current = ""
			if after, ok := strings.CutPrefix(name, "profiles."); ok {
				current = strings.Trim(after, `"`)
				if _, seen := out[current]; !seen {
					out[current] = Profile{}
				}
			}
			continue
		}
		if current == "" {
			continue
		}
		key, value, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		p := out[current]
		if err := assign(&p, strings.TrimSpace(key), strings.TrimSpace(stripComment(value))); err != nil {
			return nil, err
		}
		out[current] = p
	}
	return out, sc.Err()
}

// stripComment removes a trailing # comment, which cannot appear inside a value
// here because every value is a scalar or a string array.
func stripComment(v string) string {
	inString := false
	for i, c := range v {
		switch c {
		case '"':
			inString = !inString
		case '#':
			if !inString {
				return v[:i]
			}
		}
	}
	return v
}

func assign(p *Profile, key, raw string) error {
	switch key {
	case "key":
		p.Key = tomlArray(raw)
	case "compare":
		p.Compare = tomlArray(raw)
	case "ignore":
		p.Ignore = tomlArray(raw)
	case "trim":
		p.Trim = ptr(raw == "true")
	case "ignore_case":
		p.IgnoreCase = ptr(raw == "true")
	case "empty_is_null":
		p.EmptyIsNull = ptr(raw == "true")
	case "tolerance":
		v, err := strconv.ParseFloat(raw, 64)
		if err != nil {
			return fmt.Errorf("tolerance must be a number, got: %s", raw)
		}
		p.Tolerance = &v
	case "max_rows":
		v, err := strconv.Atoi(raw)
		if err != nil {
			return fmt.Errorf("max_rows must be an integer, got: %s", raw)
		}
		p.MaxRows = &v
	case "delimiter":
		p.Delimiter = ptr(tomlString(raw))
	case "encoding":
		p.Encoding = ptr(tomlString(raw))
	case "engine":
		p.Engine = ptr(tomlString(raw))
	}
	return nil
}

func ptr[T any](v T) *T { return &v }

func tomlString(raw string) string {
	return strings.Trim(strings.TrimSpace(raw), `"'`)
}

func tomlArray(raw string) []string {
	raw = strings.TrimSpace(raw)
	raw = strings.TrimPrefix(raw, "[")
	raw = strings.TrimSuffix(raw, "]")
	var out []string
	for _, part := range strings.Split(raw, ",") {
		if v := tomlString(part); v != "" {
			out = append(out, v)
		}
	}
	return out
}

// Apply lays a profile down under the command-line options, which then override it.
func (o *Options) Apply(p Profile) {
	if p.Key != nil {
		o.Key = p.Key
	}
	if p.Compare != nil {
		o.Compare = p.Compare
	}
	if p.Ignore != nil {
		o.Ignore = p.Ignore
	}
	if p.Trim != nil {
		o.Trim = *p.Trim
	}
	if p.IgnoreCase != nil {
		o.IgnoreCase = *p.IgnoreCase
	}
	if p.EmptyIsNull != nil {
		o.EmptyIsNull = *p.EmptyIsNull
	}
	if p.Tolerance != nil {
		o.Tolerance = *p.Tolerance
	}
	if p.MaxRows != nil {
		o.MaxRows = *p.MaxRows
	}
	if p.Delimiter != nil {
		o.Delimiter = *p.Delimiter
	}
	if p.Encoding != nil {
		o.Encoding = *p.Encoding
	}
	if p.Engine != nil {
		o.EngineName = *p.Engine
	}
}

File: rust/src/profiles.rs
//! Named comparison profiles from `csvdiff.toml`, so a recurring comparison does
//! not get retyped.
//!
//! ```toml
//! [profiles.orders]
//! key       = ["order_id", "line_no"]
//! compare   = ["qty", "price", "status"]   # omit for all common non-key columns
//! ignore    = ["updated_at"]
//! trim      = true
//! tolerance = 0.005
//! ```
//!
//! The file format is shared with the other implementations.

use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

use crate::error::{Error, Result};
use crate::options::Options;

/// One profile. Every field is optional because "absent" and "false" mean
/// different things: only the keys actually present override the defaults.
#[derive(Debug, Default, Clone, PartialEq)]
pub struct Profile {
    pub key: Option<Vec<String>>,
    pub compare: Option<Vec<String>>,
    pub ignore: Option<Vec<String>>,
    pub trim: Option<bool>,
    pub ignore_case: Option<bool>,
    pub empty_is_null: Option<bool>,
    pub tolerance: Option<f64>,
    pub max_rows: Option<usize>,
    pub delimiter: Option<char>,
    pub encoding: Option<String>,
    pub engine: Option<String>,
}

impl Profile {
    /// Lays the profile down under the command line, which then overrides it.
    pub fn apply_to(&self, opt: &mut Options) {
        if let Some(v) = &self.key {
            opt.key = v.clone();
        }
        if let Some(v) = &self.compare {
            opt.compare = Some(v.clone());
        }
        if let Some(v) = &self.ignore {
            opt.ignore = v.clone();
        }
        if let Some(v) = self.trim {
            opt.trim = v;
        }
        if let Some(v) = self.ignore_case {
            opt.ignore_case = v;
        }
        if let Some(v) = self.empty_is_null {
            opt.empty_is_null = v;
        }
        if let Some(v) = self.tolerance {
            opt.tolerance = v;
        }
        if let Some(v) = self.max_rows {
            opt.max_rows = v;
        }
        if let Some(v) = self.delimiter {
            opt.delimiter = Some(v);
        }
        if let Some(v) = &self.encoding {
            opt.encoding = v.clone();
        }
        if let Some(v) = &self.engine {
            opt.engine = v.clone();
        }
    }
}

/// Where a config file is looked for when no path is given.
pub fn search_path() -> Vec<PathBuf> {
    let mut paths = vec![PathBuf::from("csvdiff.toml")];
    if let Ok(home) = std::env::var("HOME") {
        paths.push(PathBuf::from(home).join(".config/csvdiff/csvdiff.toml"));
    }
    paths
}

/// Reads the profile tables from a config file, or returns nothing when there is none.
///
/// This reads the subset of TOML the config format actually uses — `[profiles.x]`
/// tables of scalars and string arrays — rather than pulling in a parser for a
/// file that is a dozen lines long.
pub fn load(explicit: Option<&str>) -> Result<HashMap<String, Profile>> {
    let candidates: Vec<PathBuf> = match explicit {
        Some(p) => vec![PathBuf::from(p)],
        None => search_path(),
    };
    for path in candidates {
        if !path.is_file() {
            continue;
        }
        let text = fs::read_to_string(&path)
            .map_err(|e| Error::new(format!("cannot read {}: {e}", path.display())))?;
        return parse(&text)
            .map_err(|e| Error::new(format!("cannot parse {}: {e}", path.display())));
    }
    match explicit {
        Some(p) => Err(Error::new(format!("config file not found: {p}"))),
        None => Ok(HashMap::new()),
    }
}

/// Parses the profile tables out of a config file's text.
pub fn parse(text: &str) -> Result<HashMap<String, Profile>> {
    let mut out: HashMap<String, Profile> = HashMap::new();
    let mut current: Option<String> = None;

    for raw_line in text.lines() {
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if let Some(name) = line.strip_prefix('[').and_then(|l| l.strip_suffix(']')) {
            current = name
                .trim()
                .strip_prefix("profiles.")
                .map(|n| n.trim_matches('"').to_string());
            if let Some(name) = &current {
                out.entry(name.clone()).or_default();
            }
            continue;
        }
        let Some(name) = &current else { continue };
        let Some((key, value)) = line.split_once('=') else {
            continue;
        };
        let profile = out.entry(name.clone()).or_default();
        assign(profile, key.trim(), strip_comment(value).trim())?;
    }
    Ok(out)
}

fn assign(profile: &mut Profile, key: &str, raw: &str) -> Result<()> {
    let number =
        |what: &str| -> Error { Error::new(format!("{what} must be a number, got: {raw}")) };
    match key {
        "key" => profile.key = Some(toml_array(raw)),
        "compare" => profile.compare = Some(toml_array(raw)),
        "ignore" => profile.ignore = Some(toml_array(raw)),
        "trim" => profile.trim = Some(raw == "true"),
        "ignore_case" => profile.ignore_case = Some(raw == "true"),
        "empty_is_null" => profile.empty_is_null = Some(raw == "true"),
        "tolerance" => profile.tolerance = Some(raw.parse().map_err(|_| number("tolerance"))?),
        "max_rows" => profile.max_rows = Some(raw.parse().map_err(|_| number("max_rows"))?),
        "delimiter" => profile.delimiter = toml_string(raw).chars().next(),
        "encoding" => profile.encoding = Some(toml_string(raw)),
        "engine" => profile.engine = Some(toml_string(raw)),
        _ => {}
    }
    Ok(())
}

/// Removes a trailing `#` comment, which cannot appear inside a value here
/// because every value is a scalar or a string array.
fn strip_comment(value: &str) -> &str {
    let mut in_string = false;
    for (i, c) in value.char_indices() {
        match c {
            '"' => in_string = !in_string,
            '#' if !in_string => return &value[..i],
            _ => {}
        }
    }
    value
}

fn toml_string(raw: &str) -> String {
    raw.trim().trim_matches(['"', '\'']).to_string()
}

fn toml_array(raw: &str) -> Vec<String> {
    raw.trim()
        .trim_start_matches('[')
        .trim_end_matches(']')
        .split(',')
        .map(toml_string)
        .filter(|s| !s.is_empty())
        .collect()
}
